state-synchronizer/src/tests/fuzzing.rs

```diff
@@ -0,0 +1,226 @@
+    option,
+    prelude::*,
+    strategy::Strategy,
+};
+use std::{
+    collections::HashMap,
+    num::NonZeroUsize,
+    sync::{Arc, RwLock},
+};
+
+#[test]
+fn test_state_sync_fuzzer() {
+    let mut gen = ValueGenerator::new();
+    let data = generate_corpus(&mut gen);
+    fuzzer(data);
+}
+
+pub fn generate_corpus(gen: &mut ValueGenerator) -> Vec<u8> {
+    let msg = gen.generate(state_sync_msg_strategy());
+    lcs::to_bytes(&msg).expect("[fuzzing] failed to generate corpus for state sync msg")
+}
+
+pub fn fuzzer(data: Vec<u8>) {
+    let msg = match lcs::from_bytes::<StateSynchronizerMsg>(&data) {
+        Ok(msg) => msg,
+        Err(_) => return,
+    };
+
+    // start up coordinator
+    let (_coordinator_sender, coordinator_receiver) = mpsc::unbounded();
+    let (mempool_sender, _mempool_receiver) = mpsc::channel(1_024);
+    let config = NodeConfig::default_for_validator();
+
+    let (signers, validator_info, _keys) = SynchronizerEnvHelper::initial_setup(1);
+    let genesis_li = SynchronizerEnvHelper::genesis_li(&validator_info);
+    let storage_inner = MockStorage::new(genesis_li, signers[0].clone());
+    let initial_state = storage_inner.get_local_storage_state();
+    let storage_proxy = Arc::new(RwLock::new(storage_inner));
+
+    // mock network senders
+    let (network_reqs_tx, _network_reqs_rx) =
+        libra_channel::new(QueueStyle::FIFO, NonZeroUsize::new(8).unwrap(), None);
+    let (connection_reqs_tx, _) =
+        libra_channel::new(QueueStyle::FIFO, NonZeroUsize::new(8).unwrap(), None);
+    let network_sender = StateSynchronizerSender::new(
+        PeerManagerRequestSender::new(network_reqs_tx),
+        ConnectionRequestSender::new(connection_reqs_tx),
+    );
+    let node_network_id = NodeNetworkId::new(NetworkId::Validator, 0);
+    let network_senders = vec![(node_network_id.clone(), network_sender)]
+        .into_iter()
+        .collect::<HashMap<_, _>>();
+    let mut coordinator = SyncCoordinator::new(
+        coordinator_receiver,
+        mempool_sender,
+        network_senders,
+        RoleType::Validator,
+        Waypoint::default(),
+        config.state_sync,
+        config.upstream,
+        MockExecutorProxy::new(SynchronizerEnvHelper::default_handler(), storage_proxy),
+        initial_state,
+    );
+    let mut rt = tokio::runtime::Runtime::new().unwrap();
+    rt.block_on(async move {
+        coordinator
+            .process_one_message(PeerNetworkId(node_network_id, PeerId::random()), msg)
+            .await;
+    });
+}
+
+pub fn state_sync_msg_strategy() -> impl Strategy<Value = StateSynchronizerMsg> {
+    prop_oneof![get_chunk_request_strategy(), get_chunk_response_strategy()]
+}
+
+fn get_chunk_request_strategy() -> impl Strategy<Value = StateSynchronizerMsg> {
+    (any::<GetChunkRequest>()).prop_map(|get_chunk_request| {
+        StateSynchronizerMsg::GetChunkRequest(Box::new(get_chunk_request))
+    })
+}
+
+impl Arbitrary for GetChunkRequest {
```

Comment:
should this implementation live next to where the type is defined?

Reply:
this throws a compile error because the fuzzing/testing dependency are tightly correlated, so I have to keep all the fuzzing dependencies together. That's probably also the reason why we have a separate `proptest_types` under `types`